torch/fx/experimental/Partitioner.py

```diff
@@ -63,120 +34,367 @@ class Partition:
+        self.partition_to_used_mem_bytes[partition] = used_mem_bytes
+        # Find parent partitions and child partitions for each partition.
+        self.set_parents_and_children()
+        # Combine small partitions
+        self.combine_partitions_based_on_size(available_mem_bytes)
+        # Reassign partition ids and update self.node_to_partitions.
+        self.reorganize_partitions()
+        return
+
+    def do_partition(self) -> GraphModule:
+        """Return a module with submodules (partitions)."""
         for node in self.graph_module.graph.nodes:
             if node.op == 'output':
                 break
-            node.partition_ids = self.node_to_partitions[node]
-        return
+        module_with_submodules = split_module(
+            self.graph_module,
+            self.torch_module,
+            lambda node: self.node_to_partitions[node][0]
+        )
+        return module_with_submodules
 
-    def dump_partition_DAG(self) -> DAG:
+    def dump_partition_DAG(self, module_with_submodules: GraphModule) -> DAG:
         dag = DAG()
-        for i, partition in enumerate(self.partitions):
-            input_nodes = partition.get_input_nodes()
-            output_nodes = partition.get_output_nodes()
-            parents = partition.parents
-            parent_ids = [self.partitions.index(p) for p in parents]
-            children = partition.children
-            children_ids = [self.partitions.index(c) for c in children]
-            dag.create_node(i, parent_ids, children_ids, input_nodes, output_nodes)
+        for node in module_with_submodules.graph.nodes:
+            if node.op == 'output':
+                break
+            if node.op in {'placeholder', 'get_attr'}:
+                continue
+            if node.target == operator.__getitem__:
+                continue
+            input_nodes : Dict[Node, None] = {}
+            map_arg(node.args, lambda n: input_nodes.setdefault(n))
+            map_arg(node.kwargs, lambda n: input_nodes.setdefault(n))
+            # When a node has two or more output nodes,
+            # it outputs its result to 'getitem' nodes.
+            # Those 'getitem' nodes are the output node for this node.
+            # Otherwise, the output node is this node itself.
+            if len(node.users) > 1:
+                output_nodes = list(node.users)
+            else:
+                output_nodes = [node]
+            dag.create_node(node, list(input_nodes), output_nodes)
         return dag
 
     def create_partition(self) -> Partition:
         """Create a partition and append it to self.partitions."""
         partition_id = len(self.partitions)
         assert isinstance(self.graph_module, GraphModule)
         partition = Partition(partition_id, self.graph_module)
-        self.partitions.append(partition)
+        self.partitions.add(partition)
         return partition
+
+    def combine_partitions_based_on_size(self, available_mem_bytes) -> None:
```

Comment:
Note: O(N^3) complexity

Since this can run N iterations, each of which sorts the partitions (N * log N) and calls `find_partition_to_combine_based_on_size` (N^2)

This might become an issue with very large models, for example if someone puts a large auto-generated model through the system. 

I wonder if there's some assumptions we can bake into this to reduce the complexity. I assume this would primarily act on parallel partitions. Here's how I come to my conclusion: a) Direct parent-child pairs would have been allocated to the same partition in the initial partitioning b) Any nodes with a transitive parent-child relationship would end up in the circular dependency case.

Can you check my assumptions? Would this functionality typically be used for this "horizontal fusion" case?

Also, one last thing: can you add a test case exercising a case where partitions are combined?

Reply:
Direct parent-child pair may not be allocated to the same partition in the initial partitioning. Any nodes with a transitive parent-child relationship can be found using bfs